### PR TITLE
subsys/mgmt/hawkbit: Update sample, cleanup logs, implement client-side auth, fix multipacket response handling, improve error handling

### LINKED ISSUE #37225: subsys/mgmt/hawkbit & sample: Bugs/improvements
**Describe the bug**
Found something in the `hawkbit` that might be improved while I'm debugging why the `hawkbit` sample isn't working for me:
1. ~~The [`hawkbit_init`](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L610) function is assuming the "storage" label to be with `DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL` (usually `flash0`?) whereas it can  can be with the external serial flash so this should follows where the "storage" is defined instead of a constant.~~ Fixed
2. ~~Same goes for the [`nvs_init(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL)`](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L620)~~ Fixed
3. ~~There's an extra space before the "not" [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L630) that should be removed.~~ Fixed
4. ~~For some reason it marked the image as OK [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L638), but the image still isn't OK. Which led to a reboot [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L1235) when `hawkbit_autohandler();` is invoked.~~ Expected behavior from mcuboot
5. It rebooted before the log messages get flushed, so maybe have to delay before [reboot](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L1235), if not there's no point logging since the users won't be able to read them anyway
   - or do something in the reboot function to flush the logs. (Probably out of the scope of hawkbit)
6. The `hawkbit` subsys is expecting confirmed image, but the `MCUboot` is designed to not confirm the primary image that's flashed.
7. ~~[`LOG_MODULE_REGISTER(hawkbit);`](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L12) is wrong, its verbosity can't be configured.~~ Fixed
8. [`LOG_ERR("only part 'bApp' is supported; got %s", chunk->part);`](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L512) missing `log_strdup`, throwing error in log.

**To Reproduce**
Build `hawkbit` sample with "storage" in external flash.

**Expected behavior**
`hawkbit` sample working.

**Impact**
Can't get `hawkbit` sample to work.

**Environment (please complete the following information):**
 - OS: Windows 10
 - gnuarmemb
 - v2.6.0-rc2-107-gebe282b02d9f

**Logs**
```
*** Booting Zephyr OS build v2.6.0-rc2-107-gebe282b02d9f  ***
I: Starting bootloader
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x18000
I: Jumping to the first image slot


uart:~$ *** Booting Zephyr OS build v2.6.0-rc2-107-gebe282b02d9f  ***
[00:00:02.010,000] <inf> main: Hawkbit sample app started
[00:00:04.602,000] <inf> modem_gsm: Waiting for modem to boot up
[00:00:15.308,000] <inf> modem_gsm: Modem RDY
[00:00:16.385,000] <inf> modem_gsm: PPP channel 2 connected to GSM_1
[00:00:16.691,000] <inf> modem_gsm: Manufacturer: Quectel
[00:00:16.749,000] <inf> modem_gsm: Model: EC21
[00:00:16.808,000] <inf> modem_gsm: Revision: EC21EFAR06A01M4G
[00:00:16.868,000] <inf> modem_gsm: IMEI: 867962047311785
[00:00:17.052,000] <inf> modem_gsm: Network registered, home network.
[00:00:17.061,000] <inf> modem_gsm: Attached to packet service!
[00:00:17.080,000] <inf> net_ppp: Initializing PPP to use GSM_1
[00:00:17.084,000] <inf> modem_gsm: AT channel 1 connected to GSM_2
[00:00:17.157,000] <inf> main: hawkbit_init
[00:00:18.275,000] <inf> fs_nvs: 3 Sectors of 4096 bytes
[00:00:18.275,000] <inf> fs_nvs: alloc wra: 0, fe8
[00:00:18.275,000] <inf> fs_nvs: data wra: 0, 0
[00:00:18.276,000] <inf> hawkbit: Image is not confirmed OK
[00:00:18.276,000] <inf> hawkbit: Marked image as OK
[00:00:21.953,000] <inf> hawkbit: Erased second slot
[00:00:21.953,000] <inf> main: Starting hawkbit polling mode
[00:00:22.953,000] <err> hawkbit: The current image is not confirmed
[00:00:22.953,000] <err> hawkbit: Image is unconfirmed
[00:00:22.953,000] <err> hawkbit: Rebooting to previous confirmed image.
```

It manage to print the <err> logs here because I `k_sleep()` before reboot function call. Or else the last message would be:
```
<inf> main: Starting hawkbit polling mode
```
before it repeats the bootloader process, and there's no way for anyone to immediately know what caused the reset.

## Suggested fix
1 & 2: Refer to the [`nvs` sample](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/nvs/src/main.c), determine the location of "storage" partition directly from devicetree: #37275 
```c
...
#define STORAGE_NODE DT_NODE_BY_FIXED_PARTITION_LABEL(storage)
#define FLASH_NODE DT_MTD_FROM_FIXED_PARTITION(STORAGE_NODE)
...
int hawkbit_init(void)
{
	bool image_ok;
	int ret = 0, rc = 0;
	struct flash_pages_info info;
	int32_t action_id;
	const struct device *flash_dev;

	flash_dev = DEVICE_DT_GET(FLASH_NODE);

	fs.offset = FLASH_AREA_OFFSET(storage);
	rc = flash_get_page_info_by_offs(flash_dev, fs.offset, &info);
	if (rc) {
		LOG_ERR("Unable to get storage page info");
		return -EIO;
	}

	fs.sector_size = info.size;
	fs.sector_count = 3U;

	rc = nvs_init(&fs, flash_dev->name);
...
```

3: Remove the space: #37278
```c
LOG_INF("Image is %s confirmed OK", image_ok ? "" : "not");
```

4: ~~Perform additional check if the image is actually "OK"?~~ Expected behavior from mcuboot
5: Do `k_sleep()` in `hawkbit` before the reboot command is called so that the log subsys has time to print the logs
   - (out of `hawkbit`'s scope) Implement function in the log subsys which can be called to flush all the logs
   - (out of `hawkbit`'s scope) Do `k_sleep()` in the shutdown/reboot function so that the log subsys has time to print the logs

6: I recommend this needs to be mentioned in the `hawkbit` subsys's README & update the `hawkbit` sample's README such that the user  should flash the **confirmed** image built with `CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE=y` instead of the typical one. Or else `hawkbit` will always reboot the system when `autohandler()` is invoked. (Thanks to @nvlsianpu for the enlightenment). This is especially useful for users that are new to this.

7: Should be `LOG_MODULE_REGISTER(hawkbit, CONFIG_HAWKBIT_LOG_LEVEL);` instead. #37277

8: Should be `LOG_ERR("only part 'bApp' is supported; got %s", log_strdup(chunk->part));` instead #37276

### LINKED ISSUE #37610: subsys/mgmt/hawkbit: Unable to parse json if the payload is split into 2 packets
**Describe the bug**
`response_cb` isn't able to store the complete packet from `HAWKBIT_PROBE` & `HAWKBIT_PROBE_DEPLOYMENT_BASE` properly if they were to arrive in more than a single packet.

I was getting JSON parse error and by printing the buffer passed into the function I noticed that the last 32 bytes aren't correct, which is inline with what I saw in the log: the payload arrived in 2 packets and the second part is 32 bytes.

I'm guessing that this line is causing the issue, but I don't really understand it:

```c
if ((body_data == NULL) && (final_data == HTTP_DATA_MORE)) {
```

However I understand the [`HAWKBIT_DOWNLOAD` part](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L783) and know that it should be applicable to the `HAWKBIT_PROBE` & `HAWKBIT_PROBE_DEPLOYMENT_BASE`. After applying this part solves the issue. I believe we can can use the same code to extract the body for all cases:

```c
if (hb_context.dl.http_content_size == 0) {
	body_data = rsp->body_start;
	body_len = rsp->data_len;
	/*
	 * subtract the size of the HTTP header from body_len
	 */
	body_len -= (rsp->body_start - rsp->recv_buf);
	hb_context.dl.http_content_size = rsp->content_length;
} else {
	/*
	 * more general case where body data is set, but no need
	 * to take the HTTP header into account
	 */
	body_data = rsp->body_start;
	body_len = rsp->data_len;
}

if ((rsp->body_found == 1) && (body_data == NULL)) {
	body_data = rsp->recv_buf;
	body_len = rsp->data_len;
}
```

**Logs**
[log.txt](https://github.com/zephyrproject-rtos/zephyr/files/6979638/log.txt)

### LINKED ISSUE #38137: [Coverity CID: 239555] Unchecked return value in subsys/mgmt/hawkbit/hawkbit.c
Static code scan issues found in file:

https://github.com/zephyrproject-rtos/zephyr/tree/fe2014ae1f4a7861771f86d500ac516fb5f59776/subsys/mgmt/hawkbit/hawkbit.c

Category: Error handling issues
Function: `response_cb`
Component: Other
CID: [239555](https://scan9.coverity.com/reports.htm#v29726/p12996/mergedDefectId=239555)

Details:

https://github.com/zephyrproject-rtos/zephyr/blob/fe2014ae1f4a7861771f86d500ac516fb5f59776/subsys/mgmt/hawkbit/hawkbit.c#L829

Please fix or provide comments in coverity using the link:

https://scan9.coverity.com/reports.htm#v29271/p12996.

For more information about the violation, check the [Coverity Reference](https://scan9.coverity.com/doc/en/cov_checker_ref.html#static_checker_CHECKED_RETURN). ([CWE-252](http://cwe.mitre.org/data/definitions/252.html))

Note: This issue was created automatically. Priority was set based on classification
of the file affected and the impact field in coverity. Assignees were set using the CODEOWNERS file.

### PR DESCRIPTION
**Pull request by ycsin**
This PR addresses the following issues:

- First and second commits address a few issues in the current Hawkbit sample which can confuse a new user:
   - Current sample doesn't mentioned that the image flashed using a hardware tool should be a confirmed image. Following the instructions in the current sample will result in the board rebooting indefinitely. See point 6 of #37225 
   - Current sample tests the upgrade functionality of Hawkbit using a `hello-world` which doesn't have the hawkbit subsys. So after the `hello_world` image booted up, the new image won't actually confirm itself and ping the Hawkbit server.
- Third commit is aimed to improve the consistency of the log messages and make it easier for user to identify the origin and type of error.
- Forth commit added `log_strdup` to strings from a pointer to the http response. See point 8 of #37225 
- Fifth commit is production focused and implemented Hawkbit client side DDI API authentication using `Target` or `Gateway` token.
- Sixth commit addresses http response parsing issue when the Hawkbit server sends it in multiple packets. See #37610. This implementation is basically copied from #37243
- Seventh commit workaround point 5 of #37225, so that the user is aware of what happened. Previously the Hawkbit/sample would just reboot immediately, leaving no time for the log subsys to actually print the messages.
- Eighth commit breaks the execution of code when there's an error in `response_cb` so that it can be handled properly in `hawkbit_probe`. Previously the program continues to run even if there's an error.
- Ninth commit improved the consistency of error handling past http_request in `hawkbit_probe`, and fixes a potential issue where the connection isn't cleaned up when error occured.
- Tenth commit addresses [[Coverity CID: 239555] Unchecked return value in subsys/mgmt/hawkbit/hawkbit.c](https://github.com/zephyrproject-rtos/zephyr/issues/38137)

Fixes #37610
Fixes #37225
Fixes #38137

### CODE REVIEW

**Comment by ycsin** [comment]
@sylvioalves Will you be able to review this?
@burumaj I've used your implementation in this commit: 0cb2107, I hope you don't mind.

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 1]
```diff
@@ -712,24 +724,27 @@ static void response_cb(struct http_response *rsp,
...
 		}
 
 		if (final_data == HTTP_DATA_FINAL) {
-			if (hb_context.dl.http_content_size != body_len) {
-				LOG_ERR("HTTP response len mismatch");
+			if (hb_context.dl.http_content_size != hb_context.dl.downloaded_size) {
+				LOG_ERR("HTTP response len mismatch, expected %d, got %d",
+					hb_context.dl.http_content_size,
+					hb_context.dl.downloaded_size);
 				hb_context.code_status = HAWKBIT_METADATA_ERROR;
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:737]: @Navin-Sankar I wonder for this kind of error should it just `cleanup_connection()` and `break` just like [here](https://github.com/zephyrproject-rtos/zephyr/blob/2171df5bfd55bb7897a9eb77e507d3149bafc60d/subsys/mgmt/hawkbit/hawkbit.c#L721)?

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 2]
```diff
@@ -712,24 +724,27 @@ static void response_cb(struct http_response *rsp,
...
+			hb_context.response_data[hb_context.dl.downloaded_size] = '\0';
+			ret = json_obj_parse(hb_context.response_data,
+					     hb_context.dl.downloaded_size,
 					     json_ctl_res_descr,
 					     ARRAY_SIZE(json_ctl_res_descr),
 					     &hawkbit_results.base);
 			if (ret < 0) {
-				LOG_ERR("JSON parse error");
+				LOG_ERR("JSON parse error (HAWKBIT_PROBE): %d", ret);
 				hb_context.code_status = HAWKBIT_METADATA_ERROR;
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:748]: @Navin-Sankar

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 3]
```diff
@@ -773,24 +800,25 @@ static void response_cb(struct http_response *rsp,
...
+			strncpy(hb_context.response_data + hb_context.dl.downloaded_size,
+				body_data, body_len);
+			hb_context.dl.downloaded_size += body_len;
 		}
 
 		if (final_data == HTTP_DATA_FINAL) {
-			if (hb_context.dl.http_content_size != body_len) {
+			if (hb_context.dl.http_content_size != hb_context.dl.downloaded_size) {
 				LOG_ERR("HTTP response len mismatch");
 				hb_context.code_status = HAWKBIT_METADATA_ERROR;
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:811]: @Navin-Sankar

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 4]
```diff
@@ -773,24 +800,25 @@ static void response_cb(struct http_response *rsp,
...
+			hb_context.response_data[hb_context.dl.downloaded_size] = '\0';
+			ret = json_obj_parse(hb_context.response_data,
+					     hb_context.dl.downloaded_size,
 					     json_dep_res_descr,
 					     ARRAY_SIZE(json_dep_res_descr),
 					     &hawkbit_results.dep);
 			if (ret < 0) {
-				LOG_ERR("DeploymentBase JSON parse error");
+				LOG_ERR("DeploymentBase JSON parse error: %d", ret);
 				hb_context.code_status = HAWKBIT_METADATA_ERROR;
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:822]: @Navin-Sankar

File: subsys/mgmt/hawkbit/hawkbit.c [hunk 5]
```diff
@@ -829,7 +857,7 @@ static void response_cb(struct http_response *rsp,
 			mbedtls_md_update(&hb_context.dl.hash_ctx, body_data,
 					  body_len);
 			if (ret < 0) {
-				LOG_ERR("flash write error");
+				LOG_ERR("Flash write error: %d", ret);
 				hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
```
Comment (ycsin) [inline subsys/mgmt/hawkbit/hawkbit.c:861]: @Navin-Sankar

**Comment by sylvioalves** [comment]
@ycsin is there ready for review?

**Comment by ycsin** [comment]
> @ycsin is there ready for review?

Yes, sort of. But [this](https://github.com/zephyrproject-rtos/zephyr/pull/38081#discussion_r698593808) seems like something that should be fixed.

**Comment by ycsin** [comment]
@Navin-Sankar @sylvioalves should be ready for review now.

**Comment by ycsin** [comment]
Had a force push last night because the compiler is complaining about the [cast of the header](https://github.com/zephyrproject-rtos/zephyr/pull/38081/commits/73b9883417a54a770c8026c62ee523bc1b0eaf09#diff-51dbf7bd95ecafa1efbee72465fe2ba520297cbc23472b83e6c793c1eb74d384R897), had to add `(const char **)` to appease the compiler.

Also for whoever wondered, this [line](https://github.com/zephyrproject-rtos/zephyr/pull/38081/commits/73b9883417a54a770c8026c62ee523bc1b0eaf09#diff-51dbf7bd95ecafa1efbee72465fe2ba520297cbc23472b83e6c793c1eb74d384R872) is made this way because of the buildkite error in earlier build.

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@ycsin  would be nice if you added an entry into MAINTAINERS.yml for hawkbit and it would be even better if you became the maintainer of that area :)

**Review by cfriedt — APPROVED** [review]

**Comment by ycsin** [comment]
> @ycsin would be nice if you added an entry into MAINTAINERS.yml for hawkbit and it would be even better if you became the maintainer of that area :)

Thanks for the trust and the offer, but my current understanding of these are only skin-deep, not up to the role IMO.

**Pull request merged**